### PR TITLE
Tobin desc interface api calls

### PR DESCRIPTION
**Pull request by tobine**
A series of minor updates intended to improve the DescriptorSet class and its use. The eventual aim that these changes are leading to is to closely pair the DescriptorSet API with the Vulkan API so minimal calls are made at the top level and the DescriptorSet class can then handle all the work under the covers.

### CODE REVIEW

File: layers/vk_layer_logging.h [hunk 1]
```diff
@@ -43,7 +43,7 @@ template debug_report_data *get_my_data_ptr<debug_report_data>(void *data_key,
                                                                std::unordered_map<void *, debug_report_data *> &data_map);
 
 // Utility function to handle reporting
-static inline bool debug_report_log_msg(debug_report_data *debug_data, VkFlags msgFlags, VkDebugReportObjectTypeEXT objectType,
+static inline bool debug_report_log_msg(const debug_report_data *debug_data, VkFlags msgFlags, VkDebugReportObjectTypeEXT objectType,
```
Comment (chrisforbes) [inline layers/vk_layer_logging.h]: :+1:

File: layers/core_validation.h [hunk 2]
```diff
@@ -97,12 +98,6 @@ struct MT_FB_ATTACHMENT_INFO {
     VkDeviceMemory mem;
 };
 
-struct MT_PASS_ATTACHMENT_INFO {
```
Comment (chrisforbes) [inline layers/core_validation.h]: :+1: this is just moving stuff.

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -3381,61 +3381,67 @@ static bool validateIdleDescriptorSet(const layer_data *my_data, VkDescriptorSet
...
-        if (!pDstSet->CopyUpdate(my_data->report_data, &pCDS[i], pSrcSet, &error_str)) {
-            skipCall |= log_msg(my_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_DESCRIPTOR_SET_EXT,
-                                reinterpret_cast<const uint64_t &>(pCDS[i].dstSet), __LINE__, DRAWSTATE_INVALID_UPDATE_INDEX, "DS",
-                                "vkUpdateDescriptorsSets() failed copy update from Descriptor Set 0x%" PRIx64
-                                " to Descriptor Set 0x%" PRIx64 " with error: %s",
-                                reinterpret_cast<const uint64_t &>(pCDS[i].srcSet),
-                                reinterpret_cast<const uint64_t &>(pCDS[i].dstSet), error_str.c_str());
+        auto dst_set = pCDS[i].dstSet;
+        auto src_set = pCDS[i].srcSet;
+        auto src_pair = my_data->setMap.find(src_set);
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Followup: Add a `getDescriptorSet` helper that returns the `cvdescriptorset::DescriptorSet *` or nullptr if not found, to reduce noise?

File: layers/descriptor_sets.h [hunk 4]
```diff
@@ -259,6 +259,9 @@ class BufferDescriptor : public Descriptor {
     VkDeviceSize range_;
     const std::unordered_map<VkBuffer, BUFFER_NODE> *buffer_map_;
 };
+// Helper function for Updating descriptor sets since it crosses multiple sets
+void UpdateDescriptorSets(VkDevice device, uint32_t descriptorWriteCount, const VkWriteDescriptorSet *pDescriptorWrites,
```
Comment (chrisforbes) [inline layers/descriptor_sets.h]: This commit only contains the prototype? Alright...

File: layers/parameter_validation.cpp [hunk 5]
```diff
@@ -4417,6 +4417,10 @@ VK_LAYER_EXPORT VKAPI_ATTR PFN_vkVoidFunction VKAPI_CALL vkGetDeviceProcAddr(VkD
         return (PFN_vkVoidFunction)vkResetDescriptorPool;
     if (!strcmp(funcName, "vkAllocateDescriptorSets"))
         return (PFN_vkVoidFunction)vkAllocateDescriptorSets;
+    if (!strcmp(funcName, "vkFreeDescriptorSets"))
```
Comment (chrisforbes) [inline layers/parameter_validation.cpp]: We should really just find a way to codegen this noise.

File: tests/layer_validation_tests.cpp [hunk 6]
```diff
@@ -6925,7 +6925,7 @@ TEST_F(VkLayerTest, InvalidDSUpdateStruct) {
     VkResult err;
 
     m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT,
-                                         "Unexpected UPDATE struct of type ");
+                                         ".sType must be VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET");
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: :+1:

File: layers/core_validation.cpp [hunk 7]
```diff
@@ -70,6 +70,17 @@
...
 using namespace std;
 
+// TODO : CB really needs it's own class and files so this is just temp code until that happens
+GLOBAL_CB_NODE::~GLOBAL_CB_NODE() {
+    for (uint32_t i=0; i<VK_PIPELINE_BIND_POINT_RANGE_SIZE; ++i) {
+        // Make sure that no sets hold onto deleted CB binding
+        for (auto set : lastBound[i].uniqueBoundSets) {
+            set->RemoveBoundCommandBuffer(this);
+        }
+        lastBound[i].reset();
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Wouldn't bother with this. You're just about to run the member destructors anyway.

File: layers/core_validation_types.h [hunk 8]
```diff
@@ -434,6 +433,8 @@ struct GLOBAL_CB_NODE : public BASE_NODE {
     std::vector<std::function<bool()>> validate_functions;
     std::unordered_set<VkDeviceMemory> memObjs;
     std::vector<std::function<bool(VkQueue)>> eventUpdates;
+
+    ~GLOBAL_CB_NODE();
```
Comment (chrisforbes) [inline layers/core_validation_types.h]: :+1:

**Comment by chrisforbes** [comment]
LGTM.

**Comment by tobine** [comment]
cherry-picked and pushed to master

**Pull request closed without merging**